src/zenml/cli/stack_components.py

```diff
@@ -235,22 +238,133 @@ def register_stack_component_command(
+                    f"a part of ZenML's '{flavor_wrapper.integration}' "
+                    f"integration, which is currently not installed on your "
+                    f"system. You can install it by executing: 'zenml "
+                    f"integration install {flavor_wrapper.integration}'."
+                )
+            else:
+                cli_utils.error(
+                    f"The {component_type} that you are trying to register has "
+                    f"a custom flavor '{flavor_wrapper.name}'. In order to "
+                    f"register it, ZenML needs to be able to import the flavor "
+                    f"through its source which is defined as: "
+                    f"{flavor_wrapper.source}. Unfortunately, this is not "
+                    f"possible due to the current set of available modules/"
+                    f"working directory. Please make sure that this execution "
+                    f"is carried out in an environment where this source "
+                    f"is reachable as a module."
+                )
+
     return register_stack_component_command
 
 
+def generate_stack_component_flavor_register_command(
+    component_type: StackComponentType,
+) -> Callable[[str], None]:
+    """Generates a `register` command for the flavors of a stack component."""
+
+    @click.argument(
+        "source",
+        type=str,
+        required=True,
+    )
+    def register_stack_component_flavor_command(source: str) -> None:
+        """Adds a flavor for a stack component type"""
+        cli_utils.print_active_profile()
+
+        # Check whether the module exists and is the right type
+        component_class = validate_flavor_source(
+            source=source, component_type=component_type
+        )
+
+        # Register the flavor in the given source
+        try:
+            Repository().zen_store.create_flavor(
+                name=component_class.FLAVOR,
+                stack_component_type=component_class.TYPE,
+                source=source,
+            )
+        except EntityExistsError as e:
+            cli_utils.error(str(e))
+
+    return register_stack_component_flavor_command
+
+
+def generate_stack_component_flavor_list_command(
+    component_type: StackComponentType,
+) -> Callable[[], None]:
+    """Generates a `list` command for the flavors of a stack component."""
+
+    def list_stack_component_flavor_command() -> None:
+        """Adds a flavor for a stack component type"""
+        cli_utils.print_active_profile()
+
+        from zenml.stack.registry import flavor_registry
+
+        # List all the flavors of the component type
+        zenml_flavors = [
+            f
+            for f in flavor_registry.get_flavors_by_type(
+                component_type=component_type
+            ).values()
+        ]
+
+        custom_flavors = Repository().zen_store.get_flavors_by_type(
+            component_type=component_type
```

Comment:
This reflects what I mentioned earlier about using Repository (or another class) as the facade for all flavor related operations. This convoluted logic of retrieving flavors from two sources shouldn't be here in the CLI, so close to the end user. It should be part of a dedicated class in charge of implementing all flavor management operations in a way that hides this complexity from the consumers (the CLI being just one of the consumers, but also ZenML custom user code could be one).

Reply:
Tried to reflect the answer in the comment below. Let me know if you have further questions.